src/Interpreters/Cache/SLRUFileCachePriority.h

```diff
@@ -125,7 +125,10 @@ class SLRUFileCachePriority::SLRUIterator : public IFileCachePriority::Iterator
 
     SLRUFileCachePriority * cache_priority;
     LRUFileCachePriority::LRUIterator lru_iterator;
-    const EntryPtr entry;
+    /// Entry itself is stored by lru_iterator.entry.
+    /// We have it as a separate field to use entry without requiring any lock
+    /// (which will be required if we wanted to get entry from lru_iterator.getEntry()).
+    const std::weak_ptr<Entry> entry;
```

Comment:
It seems there is no need to store entry `weak_ptr`. We hold strong reference (shared_ptr) on `lru_iterator` which holds strong reference on `entry` (`EntryPtr` via `LRUQueue::iterator`), i.e. it's guaranteed that the `entry` is valid until we explicitly somewhere reset it. So, if we don't do the last, then it'd be sufficient to hold reference on `EntryPtr` i.e. `const EntryPtr &`.

Reply:
weak_ptr seems safer, there is one tricky place where we temporarily remove queue iterator and reassign it: https://github.com/ClickHouse/ClickHouse/blob/c14d1b482f9b3e5647f688dd22e81fd6b100188b/src/Interpreters/Cache/SLRUFileCachePriority.cpp#L374-L403
Even though I tried to make it as safe as possible in terms of not loosing the iterator, I would still like to choose a safer approach with weak_ptr.